jetty-util/src/main/java/org/eclipse/jetty/util/component/ContainerLifeCycle.java

```diff
@@ -96,35 +96,62 @@ protected void doStart() throws Exception
         _doStarted = true;
 
         // start our managed and auto beans
-        for (Bean b : _beans)
+        try
         {
-            if (b._bean instanceof LifeCycle)
+            for (Bean b : _beans)
             {
-                LifeCycle l = (LifeCycle)b._bean;
-                switch(b._managed)
+                if (b._bean instanceof LifeCycle)
                 {
-                    case MANAGED:
-                        if (!l.isRunning())
-                            start(l);
-                        break;
-                        
-                    case AUTO:
-                        if (l.isRunning())
-                            unmanage(b);
-                        else
+                    LifeCycle l = (LifeCycle)b._bean;
+                    switch (b._managed)
+                    {
+                        case MANAGED:
+                            if (!l.isRunning())
+                                start(l);
+                            break;
+
+                        case AUTO:
+                            if (l.isRunning())
+                                unmanage(b);
+                            else
+                            {
+                                manage(b);
+                                start(l);
+                            }
+                            break;
+
+                        default:
+                            break;
+                    }
+                }
+            }
+
+            super.doStart();
+        }
+        catch (Throwable t)
```

Comment:
I don't think this is the right place to do this.

I would have rather done it in `AbstractLifeCycle.start()`.
There, we call `doStart()` and wrap that call in a try/catch; from there we should call `stop()`, which calls `doStop()`, which is overridden by this class and where we already have the right logic to stop all the children beans.
Note how in `doStop()` we call the beans in reverse order while here we don't.

So I really think this should be done in `AbstractLifeCycle`, not here.

Reply:
I disagree.  Doing it in `AbstractLifeCycle` would doStop the entire component and all it's children, thus the FAILED state would be lost.  For example, somebody dumping the server when catching the start exception would not see which components had failed and the stack frame might not be enough to identify which component because multiple components may have the same classes.

Doing the stop in ContainerLifeCycle means that if a failure happens in a component, any started siblings are stopped and then the exception is propagated up the tree, so started aunts and uncles can also be stopped.

Reply:
now reversed.